lib/streamlit/watcher/local_sources_watcher.py

```diff
@@ -120,60 +122,65 @@ def update_watched_modules(self):
-                # a dynamic property, which can crash if the underlying
-                # module's code has a bug (as discovered by one of our users).
-                continue
-
-        # Clone dict here because we may alter the original dict inside the
-        # loop.
-        watched_modules = dict(self._watched_modules)
-
-        # Remove no-longer-depended-on files from self._watched_modules
-        # Will this ever happen?
-        for filepath in watched_modules:
-            if filepath not in local_filepaths:
-                self._deregister_watcher(filepath)
+        modules_paths = {
+            name: self._exclude_blacklisted_paths(get_module_paths(module))
+            for name, module in dict(sys.modules).items()
+        }
+
+        self._register_necessary_watchers(modules_paths)
+
+    def _register_necessary_watchers(
+        self, module_paths: t.Dict[str, t.Set[str]]
+    ) -> None:
+        for name, paths in module_paths.items():
+            for path in paths:
+                if self._file_should_be_watched(path):
+                    self._register_watcher(path, name)
+
+    def _exclude_blacklisted_paths(self, paths: t.Set[str]) -> t.Set[str]:
+        return {p for p in paths if not self._folder_black_list.is_blacklisted(p)}
+
+
+def get_module_paths(module: types.ModuleType) -> t.Set[str]:
+    paths_extractors = [
+        lambda m: [m.__file__],
+        # https://docs.python.org/3/reference/datamodel.html
+        # __file__ is the pathname of the file from which the module was loaded
+        # if it was loaded from a file.
+        # The __file__ attribute may be missing for certain types of modules
+        lambda m: [m.__spec__.origin],
+        # https://docs.python.org/3/reference/import.html#__spec__
+        # The __spec__ attribute is set to the module spec that was used
+        # when importing the module. one exception is __main__,
+        # where __spec__ is set to None in some cases.
+        # https://www.python.org/dev/peps/pep-0451/#id16
+        # "origin" in an import context means the system
+        # (or resource within a system) from which a module originates
+        # ... It is up to the loader to decide on how to interpret
+        # and use a module's origin, if at all.
+        lambda m: [p for p in m.__path__._path]
+        # https://www.python.org/dev/peps/pep-0420/
+        # Handling of "namespace packages" in which the __path__ attribute
+        # is a _NamespacePath object with a _path attribute containing
+        # the various paths of the package.
+    ]
+
+    return _extract_module_paths_with(module, paths_extractors)
+
+
+def _extract_module_paths_with(
+    module: types.ModuleType,
+    paths_extractors: t.List[t.Callable[[types.ModuleType], t.List[t.Optional[str]]]],
+) -> t.Set[str]:
+    all_paths = set()
+    for extract_paths in paths_extractors:
+        try:
+            potential_paths = extract_paths(module)
+        except:  # AttributeErro
```

Comment:
nit: a letter got dropped at the end. I also think we may want to keep the long comment in the equivalent code that this replaces to explain why we have such a broad `except`.

Alternatively, if `AttributeErrors` are the only expected error type here, we may want to either let unexpected error types fall through (or at least log them) so that unexpected errors fail more loudly and we're able to catch bugs earlier.

Reply:
I do believe that `AttributeError` is the only thing that is reasonable to be expected here. However, in the tests a misbehaved module is created to specifically check that the watcher can handle modules that throw random exceptions. So, I thought it's best to use a broad except as before. Happy to remove it, if you think so.

In the meantime, I have added a warning in case an non-`AttributeError` is caught.

Reply:
Whoops, I must have not read the comment in the code that this replaces thoroughly enough. Looking at it again, it's pretty clear to me now that the broad except is intentional since it's possible for a module with a bug in it to throw a random exception here.

In that case, I wouldn't mind leaving the behavior as it was before this PR, but I also think that sending a warning to the user is slightly preferable.